src/System.Device.Gpio/System/Device/Gpio/Drivers/LibGpiodDriver.Linux.cs

```
@@ -0,0 +1,276 @@
+                _pinNumberToSafeLineHandle.Remove(pinNumber);
+            }
+        }
+
+        protected internal override int ConvertPinNumberToLogicalNumberingScheme(int pinNumber) => throw new
+            PlatformNotSupportedException("This driver is generic so it cannot perform conversions between pin numbering schemes.");
+
+        protected internal override PinMode GetPinMode(int pinNumber)
+        {
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+            {
+                return (Interop.GetLineDirection(pin) == 1) ? PinMode.Input : PinMode.Output;
+            }
+            return PinMode.Input;
+        }
+
+        protected internal override bool IsPinModeSupported(int pinNumber, PinMode mode)
+        {
+            // Libgpiod Api seems do not support pull up or pull down resistors.
+            return mode != PinMode.InputPullDown && mode != PinMode.InputPullUp;
+        }
+
+        protected internal override void OpenPin(int pinNumber)
+        {
+            SafeLineHandle pin = Interop.GetChipLineByOffset(_chip, pinNumber);
+            if (pin == null)
+            {
+                throw new IOException($"Pin number {pinNumber} not available for chip: {_chip} error: {Marshal.GetLastWin32Error()}");
+            }
+            _pinNumberToSafeLineHandle.Add(pinNumber, pin);
+        }
+
+        protected internal override PinValue Read(int pinNumber)
+        {
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+            {
+                int value = Interop.GetGpiodLineValue(pin);
+                if (value == -1)
+                {
+                    throw new IOException($"Error while reading value from pin number: {pinNumber}, error: {Marshal.GetLastWin32Error()}");
+                }
+                return (value == 1) ? PinValue.High : PinValue.Low;
+            }
+            return PinValue.Low;
```

Should we throw an exception here instead? basically this would mean that something went wrong since you tried to read something from a dictionary and we don't have that. Same goes for getting the PinMode and defaulting to Input.